File: libraries/python/src/urc_converter/writer.py
"""Canonical URC serializer."""

from .model import Chart, Judgment
from .strings import (
	JUDGMENT_FIELD_RATE,
	JUDGMENT_FIELD_WINDOW,
	LAYOUT_FIELD_SPECIAL,
	LAYOUT_FIELD_TYPE,
	METADATA_FIELDS,
	SECTION_JUDGMENT,
	SECTION_LAYOUT,
	SECTION_METADATA,
	SECTION_NOTES,
	SECTION_TIMING,
)


def _float(value: float) -> str:
	text = repr(value)
	if text.endswith(".0"):
		text = text[:-2]
	return text


def _judgment_lines(judgment: Judgment) -> list[str]:
	windows = ", ".join(_float(value) for value in judgment.windows)
	rates = ", ".join(_float(value) for value in judgment.rates)
	return [f"{JUDGMENT_FIELD_WINDOW}: {windows}", f"{JUDGMENT_FIELD_RATE}: {rates}"]


def write(chart: Chart) -> str:
	"""Serialize a Chart to canonical URC text."""
	layout = chart.layout
	if layout.special_keys > 0:
		type_text = f"{layout.keys}+{layout.special_keys}"
	else:
		type_text = str(layout.keys)
	if layout.special_lanes is None:
		special_text = "None"
	else:
		special_text = ", ".join(str(lane) for lane in layout.special_lanes)
	metadata = chart.metadata
	metadata_values = (
		metadata.original,
		metadata.title,
		metadata.artist,
		metadata.creator,
		metadata.version,
	)
	lines = [f"@URC {chart.format_version.major}.{chart.format_version.minor}", ""]
	lines.append(SECTION_METADATA)
	for name, value in zip(METADATA_FIELDS, metadata_values, strict=True):
		lines.append(f"{name}: {value}")
	if chart.judgment is not None:
		lines.append("")
		lines.append(SECTION_JUDGMENT)
		lines.extend(_judgment_lines(chart.judgment))
	lines.append("")
	lines.append(SECTION_LAYOUT)
	lines.append(f"{LAYOUT_FIELD_TYPE}: {type_text}")
	lines.append(f"{LAYOUT_FIELD_SPECIAL}: {special_text}")
	lines.append("")
	lines.append(SECTION_TIMING)
	for point in chart.timing:
		fields = [
			str(point.timestamp_ms),
			_float(point.bpm),
			f"{point.meter.beats}/{point.meter.note_value}",
		]
		if point.multiplier is not None:
			fields.append(_float(point.multiplier))
		lines.append(", ".join(fields))
	lines.append("")
	lines.append(SECTION_NOTES)
	for note in sorted(chart.notes, key=lambda item: (item.timestamp_ms, item.lane)):
		lines.append(f"{note.timestamp_ms}, {note.lane}, {note.type.value}")
	return "\n".join(lines) + "\n"

File: libraries/dotnet/UrcConverter/Writer.fs
namespace UrcConverter

module Writer =

    open System
    open System.Globalization

    let private floatText (value: float) =
        // Spec formatting: if integer-valued, format as integer (e.g. 1.0 -> 1)
        if value = Math.Floor value then
            (int64 value).ToString(CultureInfo.InvariantCulture)
        else
            value.ToString(CultureInfo.InvariantCulture)

    let private joined (values: float list) =
        values |> List.map floatText |> String.concat ", "

    let write (chart: Chart) : string =
        let layoutLines =
            let typeText =
                if chart.Layout.SpecialKeys > 0 then
                    $"{chart.Layout.Keys}+{chart.Layout.SpecialKeys}"
                else
                    string chart.Layout.Keys

            let specialText =
                match chart.Layout.SpecialLanes with
                | None -> Strings.specialNone
                | Some lanes -> lanes |> List.map string |> String.concat ", "

            [
                Strings.sectionLayout
                $"{Strings.layoutFieldType}: {typeText}"
                $"{Strings.layoutFieldSpecial}: {specialText}"
            ]

        let metadataValues =
            [
                chart.Metadata.Original
                chart.Metadata.Title
                chart.Metadata.Artist
                chart.Metadata.Creator
                chart.Metadata.Version
            ]

        let metadataLines =
            Strings.sectionMetadata
            :: (List.zip Strings.metadataFields metadataValues
                |> List.map (fun (name, value) -> $"{name}: {value}"))

        let judgmentLines =
            match chart.Judgment with
            | Some judgment ->
                [
                    Strings.sectionJudgment
                    $"{Strings.judgmentFieldWindow}: {joined judgment.Windows}"
                    $"{Strings.judgmentFieldRate}: {joined judgment.Rates}"
                ]
            | None -> []

        let timingLines =
            chart.TimingPoints
            |> List.map (fun point ->
                let fields =
                    [
                        string point.TimestampMs
                        floatText point.Bpm
                        $"{point.Meter.Beats}/{point.Meter.NoteValue}"
                    ]
                    @ (match point.Multiplier with
                       | Some multiplier -> [ floatText multiplier ]
                       | None -> [])

                fields |> String.concat ", ")

        let noteLines =
            chart.Notes
            |> List.sortBy (fun note -> note.TimestampMs, note.Lane)
            |> List.map (fun note -> $"{note.TimestampMs}, {note.Lane}, {note.Type.Token}")

        [
            [ $"@URC {chart.FormatVersion.Major}.{chart.FormatVersion.Minor}" ]
            metadataLines
            judgmentLines
            layoutLines
            Strings.sectionTiming :: timingLines
            Strings.sectionNotes :: noteLines
        ]
        |> List.choose (function
            | [] -> None
            | lines -> Some(String.concat "\n" lines))
        |> String.concat "\n\n"
        |> fun text -> text + "\n"
